File: lib/random.cpp
#include "number.h"

#include <random>

static std::random_device g_rd;
static std::mt19937 g_gen(g_rd());

namespace rtl {

namespace ne_random {

Number uint32()
{
    return number_from_uint32(g_gen());
}

} // namespace ne_random

} // namespace rtl

File: exec/csnex/lib/random.cs
using System;
using System.Security.Cryptography;

namespace csnex.rtl
{
    public class random
    {
        private Executor Exec;
        private RandomNumberGenerator rng;

        public random(Executor exe)
        {
            rng = RandomNumberGenerator.Create();
            Exec = exe;
        }

        public void uint32()
        {
            byte[] r = new byte[4];
            rng.GetBytes(r);
            uint val = BitConverter.ToUInt32(r, 0);

            Exec.stack.Push(new Cell(new Number(val)));
        }

        public void bytes()
        {
            uint count = Number.number_to_uint32(Exec.stack.Pop().Number);

            byte[] r = new byte[count];
            rng.GetBytes(r);

            Exec.stack.Push(new Cell(r));
        }
    }
}
